File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lyricsscraper/util/AppUtils.java
package com.wa2c.android.medoly.plugin.action.lyricsscraper.util;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.text.Html;
import android.text.TextUtils;

import com.wa2c.android.medoly.plugin.action.lyricsscraper.R;
import com.wa2c.android.medoly.plugin.action.lyricsscraper.service.ProcessService;

import java.text.Normalizer;


/**
 * App utilities.
 */
public class AppUtils {

    /**
     * Show message.
     * @param context context.
     * @param text message.
     */
    public static void showToast(Context context, String text) {
        ToastReceiver.showToast(context, text);
    }

    /**
     * Show message.
     * @param context context
     * @param stringId resource id.
     */
    public static void showToast(Context context, int stringId) {
        ToastReceiver.showToast(context, stringId);
    }



    /**
     * Get first non-null object.
     * @param objects Objects.
     * @return First non-null object. null as all null.
     */
    public static <T> T coalesce(T... objects) {
        if (objects == null)
            return null;
        for (T obj : objects) {
            if (obj != null)
                return obj;
        }
        return null;
    }

    /**
     * Get first non-null text.
     * @param texts Texts.
     * @return First non-null object. empty text as all null.
     */
    public static String coalesce(String... texts) {
        if (texts == null)
            return "";
        for (String text : texts) {
            if (!TextUtils.isEmpty(text))
                return text;
        }
        return "";
    }



    /**
     * 歌詞を調整する。
     * @param text 歌詞テキスト。
     * @return 調整後の歌詞テキスト。
     */
    @SuppressWarnings("deprecation")
    public static String adjustLyrics(String text) {
        if (TextUtils.isEmpty(text))
            return null;

        // タグ除去
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            text = Html.fromHtml(text, Html.FROM_HTML_MODE_LEGACY).toString();
        } else {
            text = Html.fromHtml(text).toString();
        }

        // トリミング
        text = trimLines(text);

        return text;
    }

    /**
     * 全角を含めてトリミング。
     * @param text 元テキスト。
     * @return トリミングテキスト。
     */
    public static String trimLines(String text) {
        if (TextUtils.isEmpty(text))
            return "";

        return text.replaceAll("(?m)^[\\t 　]*", "").replaceAll("(?m)[\\t 　]*$", "").trim();
    }



    /**
     * 比較向けにテキストをノーマライズ。
     * @param text テキスト。
     * @return 変換後テキスト。
     */
    public static String normalizeText(String text) {
        if (TextUtils.isEmpty(text))
            return "";

        // 正規化
        String output = trimLines(removeParentheses(Normalizer.normalize(text, Normalizer.Form.NFKC)).toLowerCase());
        // 特殊文字正規化
        return output
                .replace("゠", "=")
                .replace("(“|”)", "\"")
                .replace("(‘|’)", "\'");
    }

    /**
     * 括弧で括られた文字等（補助文字）を取り除く
     * @param text テキスト。
     * @return 括弧を取り除いたテキスト。
     */
    public static String removeParentheses(String text) {
        if (TextUtils.isEmpty(text))
            return "";

        return text
                .replaceAll("(^[^\\(]+)\\(.*?\\)", "$1")
                .replaceAll("(^[^\\[]+)\\[.*?\\]", "$1")
                .replaceAll("(^[^\\{]+)\\{.*?\\}", "$1")
                .replaceAll("(^[^\\<]+)\\<.*?\\>", "$1")
                .replaceAll("(^[^\\（]+)\\（.*?\\）", "$1")
                .replaceAll("(^[^\\［]+)\\［.*?\\］", "$1")
                .replaceAll("(^[^\\｛]+)\\｛.*?\\｝", "$1")
                .replaceAll("(^[^\\＜]+)\\＜.*?\\＞", "$1")
                .replaceAll("(^[^\\【]+)\\【.*?\\】", "$1")
                .replaceAll("(^[^\\〔]+)\\〔.*?\\〕", "$1")
                .replaceAll("(^[^\\〈]+)\\〈.*?\\〉", "$1")
                .replaceAll("(^[^\\《]+)\\《.*?\\》", "$1")
                .replaceAll("(^[^\\「]+)\\「.*?\\」", "$1")
                .replaceAll("(^[^\\『]+)\\『.*?\\』", "$1")
                .replaceAll("(^[^\\〖]+)\\〖.*?\\〗", "$1")
                .replaceAll("(^[^\\-]+)-.*?-", "$1")
                .replaceAll("(^[^\\－]+)－.*?－", "$1")
                .replaceAll(" (~|～|〜|〰).*", "");

    }

    /**
     * Remove text after dash characters.
     * @param text text.
     * @return removed text.
     */
    public static String removeDash(String text) {
        if (TextUtils.isEmpty(text))
            return "";

        return text
                .replaceAll("\\s+(-|－|―|ー|ｰ|~|～|〜|〰|=|＝).*", "");
    }

    /**
     * Remove attached info.
     * @param text text.
     * @return removed text.
     */
    public static String removeTextInfo(String text) {
        if (TextUtils.isEmpty(text))
            return "";

        return text
                .replaceAll("(?i)[\\(\\<\\[\\{\\s]?off vocal.*", "")
                .replaceAll("(?i)[\\(\\<\\[\\{\\s]?no vocal.*", "")
                .replaceAll("(?i)[\\(\\<\\[\\{\\s]?less vocal.*", "")
                .replaceAll("(?i)[\\(\\<\\[\\{\\s]?without.*", "")
                .replaceAll("(?i)[\\(\\<\\[\\{\\s]?w/o.*", "")
                .replaceAll("(?i)[\\(\\<\\[\\{\\s]?backtrack.*", "")
                .replaceAll("(?i)[\\(\\<\\[\\{\\s]?backing track.*", "")
                .replaceAll("(?i)[\\(\\<\\[\\{\\s]?karaoke.*", "")
                .replaceAll("(?i)[\\(\\<\\[\\{\\s]?カラオケ.*", "")
                .replaceAll("(?i)[\\(\\<\\[\\{\\s]?からおけ.*", "")
                .replaceAll("(?i)[\\(\\<\\[\\{\\s]?歌無.*", "")
                .replaceAll("(?i)[\\(\\<\\[\\{\\s]?vocal only.*", "")
                .replaceAll("(?i)[\\(\\<\\[\\{\\s]?instrumental.*", "")
                .replaceAll("(?i)[\\(\\<\\[\\{\\s]?inst\\..*", "")
                .replaceAll("(?i)[\\(\\<\\[\\{\\s]?インスト.*", "")
                ;
    }


    /**
     * 2つのテキストを比較して、ほぼ同じ場合はtrue。
     * @param text1 比較テキスト1。
     * @param text2 比較テキスト2。
     * @return ほぼ一致する場合はtrue。
     */
    public static boolean similarText(String text1, String text2) {
        if (TextUtils.isEmpty(text1) || TextUtils.isEmpty(text2))
            return false;

        String it = removeWhitespace(normalizeText(text1), false);
        String ot = removeWhitespace(normalizeText(text2), false);
        return it.equals(ot);
    }

    /**
     * 空白を置換える
     * @param text テキスト。
     * @param insertSpace スペースに置換える場合はtrue。
     * @return 変換後テキスト。
     */
    private static String removeWhitespace(String text, boolean insertSpace) {
        if (TextUtils.isEmpty(text))
            return "";

        return text.replaceAll("(\\s|　)", insertSpace ? " " : "");
    }



    /** Request code */
    public static final int REQUEST_CODE_SAVE_FILE = 1;

    /**
     * Save file.
     * @param activity A activity.
     * @param title Title (searching text).
     * @param artist Artist (searching text).
     */
    public static void saveFile(@NonNull Activity activity, String title, String artist) {
        try {
            if (title == null)
                title = "";
            if (artist == null)
                artist = "";

            SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(activity);

            String defaultNameKey = activity.getString(R.string.pref_file_name_default);
            String defaultName = pref.getString(defaultNameKey, activity.getString(R.string.file_name_default_default));

            String separatorKey = activity.getString(R.string.pref_file_name_separator);
            String separator = pref.getString(separatorKey, activity.getString(R.string.file_name_separator_default));

            String fileName;
            switch (defaultName) {
                case "TITLE_ARTIST":
                    fileName = title + separator + artist;
                    break;
                case "ARTIST_TITLE":
                    fileName = artist + separator + title;
                    break;
                default:
                    fileName = title;
                    break;
            }

            Intent intent = new Intent(Intent.ACTION_CREATE_DOCUMENT);
            intent.addCategory(Intent.CATEGORY_OPENABLE);
            intent.setType("*/*");
            intent.putExtra(Intent.EXTRA_TITLE, fileName + ".lrc");
            activity.startActivityForResult(intent, REQUEST_CODE_SAVE_FILE);
        } catch (Exception e) {
            Logger.e(e);
            showToast(activity, R.string.error_app);
        }
    }
}

File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lyricsscraper/util/AppUtils.kt
package com.wa2c.android.medoly.plugin.action.lyricsscraper.util

import android.content.Context
import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import androidx.core.text.HtmlCompat
import com.wa2c.android.medoly.library.ExtraData
import com.wa2c.android.medoly.library.MediaPluginIntent
import com.wa2c.android.medoly.library.PropertyData
import com.wa2c.android.medoly.plugin.action.lyricsscraper.R
import com.wa2c.android.prefs.Prefs
import timber.log.Timber
import java.text.Normalizer


/**
 * App utilities.
 */
object AppUtils {


    /** Request code  */
    const val REQUEST_CODE_SAVE_FILE = 1

    /**
     * Show message.
     * @param context context.
     * @param text message.
     */
    fun showToast(context: Context, text: String) {
        ToastReceiver.showToast(context, text)
    }

    /**
     * Show message.
     * @param context context
     * @param stringId resource id.
     */
    fun showToast(context: Context, stringId: Int) {
        ToastReceiver.showToast(context, stringId)
    }



    /**
     * Get first non-null object.
     * @param objects Objects.
     * @return First non-null object. null as all null.
     */
    fun <T> coalesce(vararg objects: T): T? {
        return objects.firstOrNull { it != null }
    }

    /**
     * Get first non-empty text.
     * @param texts Texts.
     * @return First non-null object. empty text as all null.
     */
    fun coalesce(vararg texts: String?): String {
        for (text in texts) {
            if (!text.isNullOrEmpty())
                return text
        }
        return ""
    }


    /**
     * Adjust lyrics. (remove tags, trim lines)
     * @param inputText lyrics text.
     * @return adjusted lyrics text.
     */
    fun adjustHtmlText(inputText: String?): String? {
        if (inputText.isNullOrEmpty() || inputText == "null") // "null" for  JOYSOUND
            return null

        // Remove tags
        var text = HtmlCompat.fromHtml(inputText, HtmlCompat.FROM_HTML_MODE_LEGACY).toString()

        // Trimming
        text = trimLines(text)

        // Convert EOL Character
        text = unifyEOLCode(text)

        return text
    }

    /**
     * Trim spaces including full-width characters.
     * @param text Text.
     * @return Trimmed text.
     */
    fun trimLines(text: String?): String {
        return if (text.isNullOrEmpty()) "" else text
                .replace("(?m)^[\\t 　]*".toRegex(), "")
                .replace("(?m)[\\t 　]*$".toRegex(), "")
                .trim { it <= ' ' }
    }

    /**
     * Unify EOL characters to CR+LF.
     * @param text Text.
     * @param eol EOL code.
     * @return Converted text.
     */
    fun unifyEOLCode(text: String?, eol: String = "\r\n"): String {
        return if (text.isNullOrEmpty()) ""
            else "\r\n|[\n\r\u2028\u2029\u0085]".toRegex().replace(text, eol)
    }

    /**
     * Normalize text.
     * @param text text.
     * @return Normalized text.
     */
    fun normalizeText(text: String?): String {
        if (text.isNullOrEmpty())
            return ""

        // normalize
        val output = trimLines(Normalizer.normalize(text, Normalizer.Form.NFKC)).toLowerCase()
        // change special characters
        return output
                .replace("゠", "=")
                .replace("(“|”)", "\"")
                .replace("(‘|’)", "\'")
    }

    /**
     * Remove parentheses.
     * @param text text.
     * @return removed text.
     */
    fun removeParentheses(text: String?): String {
        return if (text.isNullOrEmpty()) "" else text
                .replace("([^\\(]+)\\(.*?\\)".toRegex(), "$1")
                .replace("([^\\[]+)\\[.*?\\]".toRegex(), "$1")
                .replace("([^\\{]+)\\{.*?\\}".toRegex(), "$1")
                .replace("([^\\<]+)\\<.*?\\>".toRegex(), "$1")
                .replace("([^\\（]+)\\（.*?\\）".toRegex(), "$1")
                .replace("([^\\［]+)\\［.*?\\］".toRegex(), "$1")
                .replace("([^\\｛]+)\\｛.*?\\｝".toRegex(), "$1")
                .replace("([^\\＜]+)\\＜.*?\\＞".toRegex(), "$1")
                .replace("([^\\【]+)\\【.*?\\】".toRegex(), "$1")
                .replace("([^\\〔]+)\\〔.*?\\〕".toRegex(), "$1")
                .replace("([^\\〈]+)\\〈.*?\\〉".toRegex(), "$1")
                .replace("([^\\《]+)\\《.*?\\》".toRegex(), "$1")
                .replace("([^\\「]+)\\「.*?\\」".toRegex(), "$1")
                .replace("([^\\『]+)\\『.*?\\』".toRegex(), "$1")
                .replace("([^\\〖]+)\\〖.*?\\〗".toRegex(), "$1")
    }

    /**
     * Remove text after dash characters.
     * @param text text.
     * @return removed text.
     */
    fun removeDash(text: String?): String {
        return if (text.isNullOrEmpty()) "" else text
                .replace("\\s+(-|－|―|ー|ｰ|~|～|〜|〰|=|＝).*".toRegex(), "")
    }

    /**
     * Remove attached info.
     * @param text text.
     * @return removed text.
     */
    fun removeTextInfo(text: String?): String {
        return if (text.isNullOrEmpty()) "" else text
                .replace("(?i)[\\(\\<\\[\\{\\s]?off vocal.*".toRegex(), "")
                .replace("(?i)[\\(\\<\\[\\{\\s]?no vocal.*".toRegex(), "")
                .replace("(?i)[\\(\\<\\[\\{\\s]?less vocal.*".toRegex(), "")
                .replace("(?i)[\\(\\<\\[\\{\\s]?without.*".toRegex(), "")
                .replace("(?i)[\\(\\<\\[\\{\\s]?w/o.*".toRegex(), "")
                .replace("(?i)[\\(\\<\\[\\{\\s]?backtrack.*".toRegex(), "")
                .replace("(?i)[\\(\\<\\[\\{\\s]?backing track.*".toRegex(), "")
                .replace("(?i)[\\(\\<\\[\\{\\s]?karaoke.*".toRegex(), "")
                .replace("(?i)[\\(\\<\\[\\{\\s]?カラオケ.*".toRegex(), "")
                .replace("(?i)[\\(\\<\\[\\{\\s]?からおけ.*".toRegex(), "")
                .replace("(?i)[\\(\\<\\[\\{\\s]?歌無.*".toRegex(), "")
                .replace("(?i)[\\(\\<\\[\\{\\s]?vocal only.*".toRegex(), "")
                .replace("(?i)[\\(\\<\\[\\{\\s]?instrumental.*".toRegex(), "")
                .replace("(?i)[\\(\\<\\[\\{\\s]?inst\\..*".toRegex(), "")
                .replace("(?i)[\\(\\<\\[\\{\\s]?インスト.*".toRegex(), "")
    }

    /**
     * Save file.
     * @param activity A activity.
     * @param inputTitle Title (searching text).
     * @param inputArtist Artist (searching text).
     */
    fun saveFile(activity: AppCompatActivity, inputTitle: String?, inputArtist: String?) {
        var title = inputTitle
        var artist = inputArtist
        try {
            if (title == null)
                title = ""
            if (artist == null)
                artist = ""

            val prefs = Prefs(activity)
            val defaultName = prefs.getString(R.string.pref_file_name_default, defRes = R.string.pref_default_file_name_default)
            val separator = prefs.getString(R.string.pref_file_name_separator, defRes = R.string.pref_default_file_name_separator)

            val fileName  = when (defaultName) {
                "TITLE_ARTIST" -> title + separator + artist
                "ARTIST_TITLE" -> artist + separator + title
                else -> title
            }

            val intent = Intent(Intent.ACTION_CREATE_DOCUMENT)
            intent.addCategory(Intent.CATEGORY_OPENABLE)
            intent.type = "*/*"
            intent.putExtra(Intent.EXTRA_TITLE, "$fileName.txt")
            activity.startActivityForResult(intent, REQUEST_CODE_SAVE_FILE)
        } catch (e: Exception) {
            Timber.e(e)
            showToast(activity, R.string.error_app)
        }

    }

    /**
     * Send result.
     * @param context A context.
     * @param pluginIntent A plugin intent.
     * @param resultProperty A result property data.
     * @param resultExtra A result extra data.
     */
    fun sendResult(context: Context, pluginIntent: MediaPluginIntent, resultProperty: PropertyData? = null, resultExtra: ExtraData? = null) {
        context.sendBroadcast(pluginIntent.createResultIntent(resultProperty, resultExtra))
    }

}
